packages/rollup-dev-tools/src/tools/transpiler/binary-search-tree.ts

```diff
@@ -0,0 +1,355 @@
+import {
+  bytecodeToBuffer,
+  EVMBytecode,
+  Opcode,
+  EVMOpcodeAndBytes,
+  formatBytecode,
+  getPCOfEVMBytecodeIndex,
+} from '@eth-optimism/rollup-core'
+import { bufferUtils, getLogger } from '@eth-optimism/core-utils'
+import { getPUSHOpcode, getPUSHIntegerOp } from './helpers'
+import {
+  BinarySearchInternalNode,
+  BinarySearchLeafNode,
+  BinarySearchNode,
+} from '../../types/transpiler'
+
+const log = getLogger('binary-search-generator')
+
+// The max number of bytes we expect a JUMPDEST's PC to be expressible in.  Setting to 3 allows 16 MB contracts--more than enough!
+const maxBytesOfContractSize = 3
+
+// reasonTagged label for opcode PUSHing the PC of a binary search tree node to the stack
+const IS_PUSH_BINARY_SEARCH_NODE_LOCATION =
+  'IS_PUSH_BINARY_SEARCH_NODE_LOCATION'
+// reasonTagged label for JUMPDEST of a binary searchh tree node.
+const IS_BINARY_SEARCH_NODE_JUMPDEST = 'IS_BINARY_SEARCH_NODE_JUMPDEST'
+
+/**
+ * Generates a binary search tree based on a set of keys to search against, and values associated with each key.
+ *
+ * @param keys The (lowest-to-highest) ordered array of keys to be searched for.
+ * @param values The array of corresponding values to return.
+ * @returns A flat array of nodes representing the tree, with the final node in the array being the root.
+ */
+export const generateLogSearchTreeNodes = (
```

Comment:
I was a bit confused by this because a BST has a pretty standard definition, and this doesn't really build a BST. We can avoid the confusion and make this a lot more readable by changing it to
```
/**
 * Generates a binary search tree from a sorted list of values.
 * @param sortedValues The sorted T objects.
 * @returns The root node of the resulting BST.
 */
function buildBST(sortedValues: T[]): Node<T> {
  if (!sortedValues) return undefined
  const rootIndex = sortedValues.length / 2
  return {
    value: sortedValues[rootIndex]
    left: buildBST(sortedValues.slice(0, rootIndex))
    right: buildBST(sortedValues.slice(rootIndex + 1))
  }
}
```